### PR TITLE
Add in:users as a search filter to limit searches to users

### LINKED ISSUE #40287: Add `in:users` as a search filter to limit searches to users
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**
Currently when you have large installation in terms of users and you need to search for a user, you can only do that via unified search. This works and filters down the list of users in case you are at `*/settings/users` on your instance. The downside is that this then also trigger searches in all available search providers.

**Describe the solution you'd like**
Provide a app-filter `in:users` that can be added to the search query like it can be done for other apps like files and within the actual user search just remove any occurrence of `in:users` and just search with the rest of the search string (trimmed). Due to the in-filter all search providers will be skipped which is what is the desired behavior.

**Describe alternatives you've considered**
Implement an in-app search in users like present in talk but that would be way more work.

---

This shall be backported down to v25

### PR DESCRIPTION
**Pull request by sorbaugh**
* Resolves: #40287

## Summary
Add in:users filter to search bar via SearchProvider to have a mechanism to limit the search load in settings/users.

## TODO

- [x] Add new UserSearch Provider
- [x] Trim search query before executing actual search

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/settings/lib/Search/UserSearch.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2020 Joas Schilling <coding@schilljs.com>
+ *
+ * @author Joas Schilling <coding@schilljs.com>
+ * @author John Molakvoæ <skjnldsv@protonmail.com>
+ * @author Morris Jobke <hey@morrisjobke.de>
```
Comment (artonge) [inline apps/settings/lib/Search/UserSearch.php:10]: Needs to be adapted with your name

File: apps/settings/lib/Search/UserSearch.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+								IUserManager $userManager,
+								IAccountManager $accountManager,
+								IL10N $l) {
+		$this->settingsManager = $settingsManager;
+		$this->groupManager = $groupManager;
+		$this->urlGenerator = $urlGenerator;
+		$this->userManager = $userManager;
+		$this->accountManager = $accountManager;
+		$this->l = $l;
+	}
```
Comment (artonge) [inline apps/settings/lib/Search/UserSearch.php:73]: We can now simplify that by declaring the class properties in the constructor arguments like so: 
https://github.com/nextcloud/activity/pull/1285/files#diff-523ecbb0068bfbac78757e8bea247f4216c49b6508c44e0039be8ddda8d631dcR35-R37

Comment (artonge) [reply apps/settings/lib/Search/UserSearch.php:73]: As said, be careful if this needs to be backported to 25, the new syntax might not be compatible.

Comment (sorbaugh) [reply apps/settings/lib/Search/UserSearch.php:73]: Suuuper late reply but yes, since this must be backported to 25 I think we should leave like this.

**Review by artonge — COMMENTED** [review]
Tested and the code looks good.
Still need to address the copyright comment.

Enjoy the rebase !

File: apps/settings/lib/Search/UserSearch.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+								IGroupManager $groupManager,
+								IURLGenerator $urlGenerator,
+								IUserManager $userManager,
+								IAccountManager $accountManager,
+								IL10N $l) {
+		$this->settingsManager = $settingsManager;
+		$this->groupManager = $groupManager;
+		$this->urlGenerator = $urlGenerator;
+		$this->userManager = $userManager;
+		$this->accountManager = $accountManager;
```
Comment (blizzz) [inline apps/settings/lib/Search/UserSearch.php:70]: Maybe I am overseeing something, but they are all not being used?

File: apps/settings/lib/Search/UserSearch.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function search(IUser $user, ISearchQuery $query): SearchResult {
+
+		return SearchResult::complete(
+			$this->l->t('Users'),
+			[]
```
Comment (blizzz) [inline apps/settings/lib/Search/UserSearch.php:102]: Do I understand how it works correctly, that this search plugin intentionally returns an empty result set? For users page it make sense as results are shown on the content and not in the pull down. Maybe I am misinterpreting how that search works though.

The naming would block potential valid future searches for users, although there is the contacts menu for this. It's acceptable, but want to point it out.

Comment (sorbaugh) [reply apps/settings/lib/Search/UserSearch.php:102]: Yes I agree this is somewhat strange. ATM the main goal is for a customer to more efficiently search for contacts without triggering everything. You correctly pointed out that the goal is to not have a list of users shown in the pulldown. See commit where how this used to be solved (and how we want to revisit in the future)

https://github.com/nextcloud/server/pull/40413/commits/54a21d431bf097feda9eedf552be19410f0a9df1

File: apps/settings/lib/Search/UserSearch.php (hunk 3 above)
Comment (sorbaugh) [reply apps/settings/lib/Search/UserSearch.php]: Cleaned this one up, thanks for pointing it out 👍

**Review by blizzz — APPROVED** [review]
Just the copyright info ;)

**Review by AndyScherzinger — APPROVED** [review]

**Review by nfebe — APPROVED** [review]

**Review by artonge — APPROVED** [review]

**Comment by nickvergessen** [comment]
/backport ab81cd1e1e0685952d2c1818160954cbbaa9fa6b,65c70d7cc6425af0bf202d70c67eabd579d913a6,c38bafba9e13e4e735ed2d6698b6c4f5d9efc7d5,a3a599855e20b60627ac4e65fed8fdfa6b2c04f8,f66e4ee07246b285f79173c300de943b3a857bad,bdf0fe6a03beccc58a119f05d9cdb7ab643ebcfd to stable27

**Comment by blizzz** [comment]
Backportbot seems to have day off :-/ fastest to do it manually it seems

**Pull request merged**